README.md
# teller-cli ![Build Status](https://img.shields.io/travis/sebinsua/teller-cli.svg)
> Banking for your command line

This tool provides useful ways of interrogating your bank through your command line, and is not merely meant to be a one-to-one match with underlying APIs.

It uses [Teller](http://teller.io) behind-the-scenes to interact with your UK bank, so you will need to have an account there.

**:point_up_2: Want an account?** [@stevegraham can hook you up!](https://twitter.com/stevegraham)

## Usage

`teller show balance current` will show you your current account's balance.

`teller --help` for more commands.

*e.g.*

![Instructions](http://i.imgur.com/cvZRwev.png)

## Why?

#### Notifications with [`terminal-notifier`](https://github.com/julienXX/terminal-notifier)

`teller show balance current | terminal-notifier -title "Current Account Balance"`

![Notifications](http://i.imgur.com/RxCSig9.png)

#### Alert when :moneybag: low

```sh
#!/bin/sh

CURRENT_BALANCE=`teller show balance current --hide-currency`;
MIN_BALANCE=1000.00;

if (( $(bc <<< "$CURRENT_BALANCE < $MIN_BALANCE") ))
then
  echo "Your current balance has fallen below £$MIN_BALANCE" | terminal-notifier -title "💰 Alert" -subtitle "Current Balance is £$CURRENT_BALANCE";
fi
```

![Alerts](http://i.imgur.com/OXU5uyv.png)

#### :coffee: How much money do I spend at [Nanna's](http://www.nannasn1.com/)?

```
> teller list transactions current | grep "NANNA'S"
27   2015-11-12  NANNA'S             -2.70
60   2015-10-28  NANNA'S             -2.40
68   2015-10-26  NANNA'S             -5.40
101  2015-10-09  NANNA'S             -2.70
203  2015-07-17  NANNA'S             -4.60
206  2015-07-16  NANNA'S             -9.90
208  2015-07-16  NANNA'S             -9.30
209  2015-07-16  NANNA'S             -0.10
```

Hopefully Teller will add support for querying transactions soon.

#### Am I saving money with a chart :chart_with_upwards_trend: with [`spark`](https://github.com/holman/spark)

```
> teller list balances business --interval=monthly --timeframe=year --output=spark | spark
▁▁▁▂▃▂▃▄▄▅▆█
```

#### Have I spent more money this month than I normally do?

```sh
#!/bin/sh

CURRENT_OUTGOING=`teller show outgoing current --hide-currency | sed 's/^-//'`;
OUTGOINGS=`teller list outgoings current --output=spark`;
SUM_OUTGOING=`echo "$OUTGOINGS" | sed 's/ /+/g' | bc -l | sed 's/^-//'`;
COUNT_OUTGOING=`echo "$OUTGOINGS" | wc -w | xargs`;
AVERAGE_OUTGOING=`bc <<< "scale=2; $SUM_OUTGOING / $COUNT_OUTGOING"`;

if (( $(bc <<< "$CURRENT_OUTGOING > $AVERAGE_OUTGOING") ))
then
  DIFFERENCE_OUTGOING=`bc <<< "scale=2; $CURRENT_OUTGOING - $AVERAGE_OUTGOING"`;
  echo "You've spent £$DIFFERENCE_OUTGOING more than normal." | terminal-notifier -title "💰 Spending Alert" -subtitle "Current Outgoing is £$CURRENT_OUTGOING";
fi
```

#### Keep track of your spending from the OSX Menu Bar with [BitBar](https://github.com/matryer/bitbar)

Create a [`track-spending.1h.sh`](https://github.com/matryer/bitbar-plugins/blob/master/Finance/teller-track-spending.1h.sh) within your plugins directory:
```sh
#!/bin/sh
#
# Teller.io Banking via the OSX menu bar
# Requires:
# - a Teller.io account
# - a UK Bank
# - teller-cli: https://github.com/sebinsua/teller-cli#from-release
# - pcregrep: `brew install pcre`
#
# <bitbar.title>teller-track-spending</bitbar.title>
# <bitbar.version>v1.3.1</bitbar.version>
# <bitbar.author>Seb Insua</bitbar.author>
# <bitbar.author.github>sebinsua</bitbar.author.github>
# <bitbar.desc>Track your spending from your menu bar</bitbar.desc>
# <bitbar.image>https://camo.githubusercontent.com/e0215e6736172334f62effff36ff8df1ab38fed1/687474703a2f2f692e696d6775722e636f6d2f627638545a4c652e706e67</bitbar.image>
# <bitbar.dependencies>teller-cli, pcregrep</bitbar.dependencies>
# <bitbar.abouturl>https://github.com/sebinsua/teller-cli</bitbar.abouturl>

export PATH="/usr/local/bin:/usr/bin/:$PATH";

SPENDING_LIMIT='3000.00'; # Change this to a suitable spending limit.

exit_if_zero() {
  RETURN_CODE=$1;
  ERROR_MESSAGE=$2;
  if [ "$ERROR_MESSAGE" = "" ]; then
    ERROR_MESSAGE="Offline";
  fi;
  if [ "$RETURN_CODE" -ne 0 ]; then
    echo "$ERROR_MESSAGE|color=#7e7e7e";
    exit 1;
  fi;
}

# If we're offline we shouldn't output junk in the menu bar.
curl --connect-timeout 5 www.google.com > /dev/null 2> /dev/null;
exit_if_zero $? "Offline";

CURRENT_OUTGOING=$(teller show outgoing current --hide-currency);
exit_if_zero $? "Error";

CURRENT_BALANCE=$(teller show balance current --hide-currency);
exit_if_zero $? "Error";

LAST_TRANSACTION=$(teller list transactions | tail -n 1 | pcregrep -o1 "[0-9]+[ ]+(.*)");
exit_if_zero $? "Error";

if [ "$(echo "$CURRENT_OUTGOING > $SPENDING_LIMIT" | bc)" -ne 0 ]; then
  OVERSPEND=$(echo "scale=2; $CURRENT_OUTGOING - $SPENDING_LIMIT" | bc);
  echo "🚨 £$OVERSPEND OVERSPENT|color=red";
else
  UNDERSPEND=$(echo "scale=2; $SPENDING_LIMIT - $CURRENT_OUTGOING" | bc);
  if [ "$(echo "$UNDERSPEND > ($SPENDING_LIMIT/2)" | bc)" -ne 0 ]; then
    echo "🏦 £$UNDERSPEND remaining|color=green";
  else
    echo "🏦 £$UNDERSPEND remaining|color=#ffbf00";
  fi;
fi;
echo "---";
echo "Current Account: £$CURRENT_BALANCE";
echo "Current Outgoing: £$CURRENT_OUTGOING";
echo "Last TX: $LAST_TRANSACTION";
```

![Tracking spending in the OSX Menu Bar](http://i.imgur.com/bv8TZLe.png)

## Installation

### From release

```
> curl -L https://github.com/sebinsua/teller-cli/releases/download/v0.0.7/teller > /usr/local/bin/teller && chmod +x /usr/local/bin/teller
```

### From source

First `git clone` and then:

```
> cargo build --release && cp ./target/release/teller /usr/local/bin && chmod +x /usr/local/bin/teller
```

## FAQ

#### Compiling gives `openssl/hmac.h` not found error

Ensure that both [Homebrew](https://github.com/Homebrew/homebrew) and `openssl` are installed, and then [try running `brew link --force openssl`](https://github.com/sfackler/rust-openssl/issues/255).

This relates to the following error:

```
--- stderr
src/openssl_shim.c:1:10: fatal error: 'openssl/hmac.h' file not found
#include <openssl/hmac.h>
```


CHANGELOG.md
# Change Log

All notable changes to this project will be [documented](http://keepachangelog.com/) in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [v0.0.7](https://github.com/sebinsua/teller-cli/releases/tag/v0.0.7) - 2016-01-29

- Unit tests have been written.
- Code has been refactored.

## [v0.0.4](https://github.com/sebinsua/teller-cli/releases/tag/v0.0.4) - 2015-12-25

- `list counterparties` gives access to a list of outgoing transactions grouped by their counterparty.
- Outgoing values are now positive since the word outgoing already describes a negative flow of money.

## [v0.0.3](https://github.com/sebinsua/teller-cli/releases/tag/v0.0.3) - 2015-12-22

- `list totals` became `list balances|outgoing|incoming`.
- `show outgoing` and `show incoming` now fetch the amounts of coming money out or into a bank account for the current month.

## [v0.0.2](https://github.com/sebinsua/teller-cli/releases/tag/v0.0.2) - 2015-12-22

- `list balances` became `list totals` and now supports `--amount-type=incoming|outgoing|all`.

## [v0.0.1](https://github.com/sebinsua/teller-cli/releases/tag/v0.0.1) - 2015-12-21

- Preliminary release.


src/api/client.rs
use hyper::{Client, Url};
use hyper::header::{Authorization, Bearer};
use rustc_serialize::json;
use chrono::{Date, DateTime, UTC};

use std::io::prelude::*; // Required for read_to_string use later.

use api::error::TellerClientError;

pub type ApiServiceResult<T> = Result<T, TellerClientError>;

#[derive(Debug, RustcDecodable)]
struct AccountResponse {
    data: Account,
}

#[derive(Debug, RustcDecodable)]
struct AccountsResponse {
    data: Vec<Account>,
}

#[derive(Debug, RustcDecodable)]
struct TransactionsResponse {
    data: Vec<Transaction>,
}

#[derive(Debug, RustcDecodable, Clone)]
pub struct Account {
    pub name: String,
    pub institution: String,
    pub id: String,
    pub currency: String,
    pub bank_code: String,
    pub balance: String,
    pub account_number_last_4: String,
}

#[derive(Debug, RustcDecodable, Clone)]
pub struct Transaction {
    pub description: String,
    pub date: String,
    pub counterparty: String,
    pub amount: String,
}

pub fn parse_utc_date_from_transaction(t: &Transaction) -> Date<UTC> {
    generate_utc_date_from_date_str(&t.date)
}

pub fn generate_utc_date_from_date_str(d: &str) -> Date<UTC> {
    let full_date = &(d.to_owned() + "T00:00:00-00:00");
    let date_without_tz = DateTime::parse_from_rfc3339(full_date).unwrap().date();
    let date = date_without_tz.with_timezone(&UTC);
    date
}

const TELLER_API_SERVER_URL: &'static str = "https://api.teller.io";

pub struct TellerClient<'a> {
    client: Client,
    auth_token: &'a str,
}

impl<'a> TellerClient<'a> {
    pub fn new(auth_token: &'a str) -> TellerClient {
        let client = Client::new();
        TellerClient {
            client: client,
            auth_token: auth_token,
        }
    }

    #[allow(dead_code)]
    pub fn new_with_hyper_client(auth_token: &'a str, client: Client) -> TellerClient {
        TellerClient {
            client: client,
            auth_token: auth_token,
        }
    }

    fn get_body(&self, url: &str) -> ApiServiceResult<String> {
        let mut res = try!(self.client.get(url)
                               .header(Authorization(
                                   Bearer { token: self.auth_token.to_string() }
                               )).send());
        if res.status.is_client_error() {
            return Err(TellerClientError::AuthenticationError);
        }

        let mut body = String::new();
        try!(res.read_to_string(&mut body));

        info!("GET {}", url);
        debug!("Response: {}", body);

        Ok(body)
    }

    pub fn get_accounts(&self) -> ApiServiceResult<Vec<Account>> {
        let body = try!(self.get_body(&format!("{}/accounts", TELLER_API_SERVER_URL)));
        let accounts_response: AccountsResponse = try!(json::decode(&body));

        Ok(accounts_response.data)
    }

    pub fn get_account(&self, account_id: &str) -> ApiServiceResult<Account> {
        let body = try!(self.get_body(&format!("{}/accounts/{}", TELLER_API_SERVER_URL, account_id)));
        let account_response: AccountResponse = try!(json::decode(&body));

        Ok(account_response.data)
    }

    pub fn raw_transactions(&self,
                            account_id: &str,
                            page_size: u32,
                            page: u32)
                            -> ApiServiceResult<Vec<Transaction>> {
        let mut url = Url::parse(&format!("{}/accounts/{}/transactions",
                                          TELLER_API_SERVER_URL,
                                          account_id)).unwrap();

        const PAGE_SIZE: &'static str = "page_size";
        const PAGE: &'static str = "page";
        let query = vec![(PAGE_SIZE, page_size.to_string()), (PAGE, page.to_string())];
        url.set_query_from_pairs(query.into_iter());

        let body = try!(self.get_body(&url.serialize()));
        let transactions_response: TransactionsResponse = try!(json::decode(&body));

        Ok(transactions_response.data)
    }

    #[allow(unused_variables)]
    pub fn get_transactions(&self,
                            account_id: &str,
                            from: &Date<UTC>,
                            to: &Date<UTC>)
                            -> ApiServiceResult<Vec<Transaction>> {
        let page_through_transactions = |from| -> ApiServiceResult<Vec<Transaction>> {
            let mut all_transactions = vec![];

            let mut fetching = true;
            let mut page = 1;
            let page_size = 250;
            while fetching {
                let mut transactions = try!(self.raw_transactions(&account_id, page_size, page));
                match transactions.last() {
                    None => {
                        // If there are no transactions left, do not fetch forever...
                        fetching = false
                    }
                    Some(past_transaction) => {
                        let past_transaction_date = parse_utc_date_from_transaction(&past_transaction);
                        if past_transaction_date <= from {
                            fetching = false;
                        }
                    }
                };

                all_transactions.append(&mut transactions);
                page = page + 1;
            }

            all_transactions = all_transactions.into_iter()
                                               .filter(|t| {
                                                   let transaction_date =
                                                       parse_utc_date_from_transaction(&t);
                                                   transaction_date >= from
                                               })
                                               .collect();

            all_transactions.reverse();
            Ok(all_transactions)
        };

        page_through_transactions(*from)
    }

}

#[cfg(test)]
mod tests {
    use super::{TellerClient, Account, Transaction, generate_utc_date_from_date_str};

    use std::error::Error;
    use hyper;

    mock_connector!(FailAuthenticationRequest {
        "https://api.teller.io" => include_str!("./mocks/fail-authentication.http")
    });
    mock_connector!(GetAccountRequest {
        "https://api.teller.io" => include_str!("./mocks/get-account.http")
    });
    mock_connector!(GetAccountsRequest {
        "https://api.teller.io" => include_str!("./mocks/get-accounts.http")
    });
    mock_connector!(GetTransactionsRequest {
        "https://api.teller.io" => include_str!("./mocks/get-transactions.http")
    });

    #[test]
    #[allow(unused)]
    fn can_instantiate_teller_client() {
        let client = TellerClient::new("auth_token");
        assert!(true);
    }

    #[test]
    #[allow(unused)]
    fn can_instantiate_account() {
        let account = Account {
            name: "Current Account".to_string(),
            institution: "Natwest".to_string(),
            id: "some-long-uuid".to_string(),
            currency: "GBP".to_string(),
            bank_code: "00000000".to_string(),
            balance: "1000.00".to_string(),
            account_number_last_4: "0000".to_string(),
        };
        assert!(true);
    }

    #[test]
    #[allow(unused)]
    fn can_instantiate_transaction() {
        let transaction = Transaction {
            description: "4836 19JAN16 C , NANNA'S , LONDON GB".to_string(),
            date: "2016-01-21".to_string(),
            counterparty: "NANNA'S".to_string(),
            amount: "-10.00".to_string(),
        };
        assert!(true);
    }

    #[test]
    fn can_fail_authentication() {
        let c = hyper::client::Client::with_connector(FailAuthenticationRequest::default());
        let client = TellerClient::new_with_hyper_client("fake-auth-token", c);
        let get_account_state = client.get_account("123");

        assert_eq!(true, get_account_state.is_err());

        let auth_err = get_account_state.unwrap_err();
        assert_eq!("Could not authenticate", auth_err.description());
    }

    #[test]
    fn can_get_account() {
        let c = hyper::client::Client::with_connector(GetAccountRequest::default());
        let client = TellerClient::new_with_hyper_client("fake-auth-token", c);
        let account = client.get_account("123").unwrap();

        assert_eq!("123", account.id);
        assert_eq!("natwest", account.institution);
        assert_eq!("current", account.name);
        assert_eq!("1000.00", account.balance);
        assert_eq!("GBP", account.currency);
        assert_eq!("000000", account.bank_code);
        assert_eq!("0000", account.account_number_last_4);
    }

    #[test]
    fn can_get_accounts() {
        let c = hyper::client::Client::with_connector(GetAccountsRequest::default());
        let client = TellerClient::new_with_hyper_client("fake-auth-token", c);
        let accounts = client.get_accounts().unwrap();

        assert_eq!("Savings", accounts[0].name);
        assert_eq!("Current", accounts[1].name);
    }

    #[test]
    fn can_get_transactions() {
        let c = hyper::client::Client::with_connector(GetTransactionsRequest::default());
        let client = TellerClient::new_with_hyper_client("fake-auth-token", c);

        let from = generate_utc_date_from_date_str("2015-01-01");
        let to = generate_utc_date_from_date_str("2016-01-01");
        let transactions = client.get_transactions("123", &from, &to).unwrap();

        assert_eq!("COUNTERPARTY-1", transactions[9].counterparty);
        assert_eq!("COUNTERPARTY-2", transactions[8].counterparty);
    }

}
